src/main/java/org/opensearch/ad/task/ADTaskManager.java

```diff
@@ -405,6 +517,235 @@ protected void forwardStaleRunningEntitiesToCoordinatingNode(
             );
     }
 
+    /**
+     * Check available task slots before start historical analysis and scale task lane.
+     * This check will be done on lead node which will gather detector task slots of all
+     * data nodes and calculate how many task slots available.
+     *
+     * @param adTask AD task
+     * @param detector detector
+     * @param detectionDateRange detection date range
+     * @param user user
+     * @param targetActionOfTaskSlotChecking target task action to run after task slot checking
+     * @param transportService transport service
+     * @param listener action listener
+     */
+    public void checkTaskSlots(
+        ADTask adTask,
+        AnomalyDetector detector,
+        DetectionDateRange detectionDateRange,
+        User user,
+        ADTaskAction targetActionOfTaskSlotChecking,
+        TransportService transportService,
+        ActionListener<AnomalyDetectorJobResponse> listener
+    ) {
+        String detectorId = detector.getDetectorId();
+        logger.debug("Start checking task slots for detector: {}, task action: {}", detectorId, targetActionOfTaskSlotChecking);
+        if (!checkingTaskSlot.tryAcquire()) {
+            logger.info("Can't acquire checking task slot semaphore for detector {}", detectorId);
+            listener
+                .onFailure(
+                    new OpenSearchStatusException(
+                        "Too many historical analysis requests in short time. Please retry later.",
+                        RestStatus.FORBIDDEN
+                    )
+                );
+            return;
+        }
+        ActionListener<AnomalyDetectorJobResponse> wrappedActionListener = ActionListener.runAfter(listener, () -> {
+            checkingTaskSlot.release(1);
+            logger.debug("Release checking task slot semaphore on lead node for detector {}", detectorId);
+        });
+        hashRing.getNodesWithSameLocalAdVersion(nodes -> {
+            int maxAdTaskSlots = nodes.length * maxAdBatchTaskPerNode;
+            ADStatsRequest adStatsRequest = new ADStatsRequest(nodes);
+            adStatsRequest
+                .addAll(ImmutableSet.of(AD_USED_BATCH_TASK_SLOT_COUNT.getName(), AD_DETECTOR_ASSIGNED_BATCH_TASK_SLOT_COUNT.getName()));
+            client.execute(ADStatsNodesAction.INSTANCE, adStatsRequest, ActionListener.wrap(adStatsResponse -> {
+                int totalUsedTaskSlots = 0; // Total entity tasks running on worker nodes
+                int totalAssignedTaskSlots = 0; // Total assigned task slots on coordinating nodes
```

Comment:
May be helpful to add some comments on the difference and meaning of each of these types of task slots